language-support/hs/bindings/src/DA/Ledger/Services/TransactionService.hs

```diff
@@ -30,112 +31,89 @@ getTransactions req =
     makeLedgerService $ \timeout config -> do
     stream <- newStream
     _ <- forkIO $
-        withGRPCClient config $ \client -> do
+        HL.withGRPCClient config $ \client -> do
             service <- LL.transactionServiceClient client
             let LL.TransactionService {transactionServiceGetTransactions=rpc} = service
-            sendToStreamFlat timeout (lowerRequest req) f stream rpc
+            GRPC.sendToStreamFlat timeout (lowerRequest req) f stream rpc
     return stream
-    where f = raiseList raiseTransaction . LL.getTransactionsResponseTransactions
+    where f = Convert.raiseList Convert.raiseTransaction . LL.getTransactionsResponseTransactions
 
 getTransactionTrees :: GetTransactionsRequest -> LedgerService (Stream TransactionTree)
 getTransactionTrees req =
     makeLedgerService $ \timeout config -> do
     stream <- newStream
     _ <- forkIO $
-        withGRPCClient config $ \client -> do
+        HL.withGRPCClient config $ \client -> do
             service <- LL.transactionServiceClient client
             let LL.TransactionService {transactionServiceGetTransactionTrees=rpc} = service
-            sendToStreamFlat timeout (lowerRequest req) f stream rpc
+            GRPC.sendToStreamFlat timeout (lowerRequest req) f stream rpc
     return stream
-    where f = raiseList raiseTransactionTree . LL.getTransactionTreesResponseTransactions
+    where f = Convert.raiseList Convert.raiseTransactionTree . LL.getTransactionTreesResponseTransactions
+
+getTransactionBy :: (LL.TransactionService HL.ClientRequest HL.ClientResult
+                        -> HL.ClientRequest 'HL.Normal request response
+                        -> IO (HL.ClientResult stream response))
+                    -> (LedgerId -> tx_ref -> [Party] -> request)
+                    -> (response -> Convert.Perhaps tx)
+                    -> (LedgerId -> tx_ref -> [Party] -> LedgerService (Maybe tx))
+getTransactionBy get_rpc mk_request resp_handler
+                 ledger_id transaction_ref parties
+    = makeLedgerService $ \timeout config -> do
+      HL.withGRPCClient config $ \client -> do
+          service <- LL.transactionServiceClient client
+          let rpc = get_rpc service
+          rpc (HL.ClientNormalRequest (mk_request ledger_id transaction_ref parties) timeout GRPC.emptyMdm)
+          >>= \case
+              HL.ClientNormalResponse resp _m1 _m2 _status _details -> case resp_handler resp of
+                                                                         Left reason -> fail (show reason)
+                                                                         Right x -> return $ Just x
+              HL.ClientErrorResponse (HL.ClientIOError (HL.GRPCIOBadStatusCode HL.StatusNotFound _details)) ->
+                  return Nothing
+              HL.ClientErrorResponse e ->
+                  fail (show e)
+              HL.ClientWriterResponse _ _ _ _ _ -> fail ""
```

Comment:
Did the typechecker previously figure out that these 3 cases were not possible? We should either try to convince it that this is still the case which probably involves adding some more constraints or if that’s not (easily) possible, emit a more useful error message.

Reply:
Yes it did. It's caused because of the abstraction of `get_rpc`, which is my opinion is not a good idea.